File: competitive-ai-frameworks/frameworks/user-flows/coordinator.py
#!/usr/bin/env python3
"""
User Flow Olympics Coordinator

Three teams compete to optimize user experience flows:
- Team 1: Happy Path Optimizers
- Team 2: Edge Case Handlers
- Team 3: Integration Specialists
"""

import argparse
import json
from pathlib import Path
from typing import Dict, List, Optional
from dataclasses import dataclass
from datetime import datetime


@dataclass
class FlowMetrics:
    """User flow performance metrics"""
    # Core metrics
    completion_rate: float = 0.0  # Percentage of successful completions
    average_time: float = 0.0     # Average completion time (seconds)
    error_rate: float = 0.0       # Percentage of errors encountered

    # UX metrics
    friction_points: int = 0       # Number of UX friction points
    accessibility_score: float = 0.0  # Accessibility compliance
    mobile_score: float = 0.0      # Mobile responsiveness

    # Integration metrics
    api_reliability: float = 0.0   # API success rate
    state_consistency: float = 0.0  # State management score
    cross_flow_integration: float = 0.0  # Integration with other flows

    def total_score(self) -> float:
        """Calculate weighted total score"""
        return (
            self.completion_rate * 0.30 +
            (100 - self.average_time) * 0.10 +  # Lower time is better
            (100 - self.error_rate) * 0.15 +     # Lower errors is better
            (100 - self.friction_points * 5) * 0.10 +  # Fewer friction points
            self.accessibility_score * 0.10 +
            self.mobile_score * 0.10 +
            self.api_reliability * 0.10 +
            self.state_consistency * 0.05
        )


@dataclass
class UserFlow:
    """Represents a user flow to test"""
    name: str
    steps: List[str]
    critical: bool = True


class UserFlowCoordinator:
    """Coordinates user flow testing championship"""

    def __init__(
        self,
        flows: List[str],
        rounds: int = 4,
        target_path: Optional[Path] = None
    ):
        self.flows = self._parse_flows(flows)
        self.rounds = rounds
        self.target_path = target_path
        self.teams = {
            "team1": "Happy Path Optimizers",
            "team2": "Edge Case Handlers",
            "team3": "Integration Specialists"
        }

    def _parse_flows(self, flow_names: List[str]) -> List[UserFlow]:
        """Parse flow names into UserFlow objects"""
        # Common user flows
        flow_definitions = {
            "registration": UserFlow(
                name="User Registration",
                steps=["Landing", "Form", "Validation", "Confirmation"],
                critical=True
            ),
            "login": UserFlow(
                name="User Login",
                steps=["Login Page", "Auth", "Dashboard"],
                critical=True
            ),
            "checkout": UserFlow(
                name="Checkout Process",
                steps=["Cart", "Shipping", "Payment", "Confirmation"],
                critical=True
            ),
            "profile": UserFlow(
                name="Profile Management",
                steps=["View Profile", "Edit", "Save", "Verify"],
                critical=False
            ),
            "search": UserFlow(
                name="Search & Browse",
                steps=["Search", "Filter", "View Results", "Select"],
                critical=False
            )
        }

        return [
            flow_definitions.get(name.lower(), UserFlow(name=name, steps=[]))
            for name in flow_names
        ]

    def run_championship(self) -> Dict:
        """Run full championship"""
        print(f"\n{'='*60}")
        print("  USER FLOW OLYMPICS")
        print(f"{'='*60}\n")
        print(f"Testing {len(self.flows)} flows:")
        for flow in self.flows:
            print(f"  - {flow.name}")
        print()

        results = {
            "flows": [f.name for f in self.flows],
            "rounds": [],
            "final_scores": {}
        }

        baseline = self._measure_baseline()

        for round_num in range(1, self.rounds + 1):
            print(f"\nRound {round_num}/{self.rounds}")
            print("─" * 60)

            round_results = self._run_round(round_num, baseline)
            results["rounds"].append(round_results)

            self._display_round_results(round_results)

        # Determine winner
        winner = self._determine_winner(results)
        results["winner"] = winner

        self._display_final_results(results)

        return results

    def _measure_baseline(self) -> Dict[str, FlowMetrics]:
        """Measure baseline flow performance"""
        print("Measuring baseline flow performance...\n")

        baseline = {}
        for flow in self.flows:
            # Simulate baseline measurements
            baseline[flow.name] = FlowMetrics(
                completion_rate=75.0,
                average_time=45.0,
                error_rate=15.0,
                friction_points=5,
                accessibility_score=60.0,
                mobile_score=70.0,
                api_reliability=85.0,
                state_consistency=80.0
            )

        return baseline

    def _run_round(
        self,
        round_num: int,
        baseline: Dict[str, FlowMetrics]
    ) -> Dict:
        """Execute one round of testing"""
        results = {}

        for team_id, team_name in self.teams.items():
            print(f"  ▶ {team_name} testing...")

            # Simulate flow optimizations
            optimizations = self._simulate_team_optimizations(
                team_id,
                baseline
            )

            score = self._calculate_improvement_score(baseline, optimizations)

            results[team_id] = {
                "team_name": team_name,
                "optimizations": optimizations,
                "score": score
            }

        return results

    def _simulate_team_optimizations(
        self,
        team_id: str,
        baseline: Dict[str, FlowMetrics]
    ) -> Dict[str, FlowMetrics]:
        """Simulate team optimizations"""

        optimizations = {}

        for flow_name, base_metrics in baseline.items():
            if team_id == "team1":  # Happy Path Optimizers
                optimizations[flow_name] = FlowMetrics(
                    completion_rate=base_metrics.completion_rate + 15,
                    average_time=base_metrics.average_time - 10,
                    error_rate=base_metrics.error_rate - 3,
                    friction_points=base_metrics.friction_points - 3,
                    accessibility_score=base_metrics.accessibility_score + 5,
                    mobile_score=base_metrics.mobile_score + 5,
                    api_reliability=base_metrics.api_reliability + 3,
                    state_consistency=base_metrics.state_consistency + 5
                )
            elif team_id == "team2":  # Edge Case Handlers
                optimizations[flow_name] = FlowMetrics(
                    completion_rate=base_metrics.completion_rate + 8,
                    average_time=base_metrics.average_time - 2,
                    error_rate=base_metrics.error_rate - 12,
                    friction_points=base_metrics.friction_points - 1,
                    accessibility_score=base_metrics.accessibility_score + 25,
                    mobile_score=base_metrics.mobile_score + 20,
                    api_reliability=base_metrics.api_reliability + 5,
                    state_consistency=base_metrics.state_consistency + 3
                )
            else:  # Integration Specialists
                optimizations[flow_name] = FlowMetrics(
                    completion_rate=base_metrics.completion_rate + 10,
                    average_time=base_metrics.average_time - 5,
                    error_rate=base_metrics.error_rate - 8,
                    friction_points=base_metrics.friction_points - 2,
                    accessibility_score=base_metrics.accessibility_score + 8,
                    mobile_score=base_metrics.mobile_score + 8,
                    api_reliability=base_metrics.api_reliability + 12,
                    state_consistency=base_metrics.state_consistency + 15
                )

        return optimizations

    def _calculate_improvement_score(
        self,
        baseline: Dict[str, FlowMetrics],
        optimized: Dict[str, FlowMetrics]
    ) -> float:
        """Calculate improvement score"""
        total_improvement = 0.0

        for flow_name in baseline.keys():
            base_score = baseline[flow_name].total_score()
            opt_score = optimized[flow_name].total_score()
            total_improvement += (opt_score - base_score)

        return total_improvement

    def _display_round_results(self, results: Dict):
        """Display round results"""
        sorted_teams = sorted(
            results.items(),
            key=lambda x: x[1]["score"],
            reverse=True
        )

        for rank, (team_id, data) in enumerate(sorted_teams, 1):
            medal = ["🥇", "🥈", "🥉"][rank-1] if rank <= 3 else "  "
            print(f"  {medal} {data['team_name']:<35} +{data['score']:.1f} pts")

    def _determine_winner(self, results: Dict) -> Dict:
        """Determine overall winner"""
        team_totals = {
            team_id: sum(
                round_data[team_id]["score"]
                for round_data in results["rounds"]
            )
            for team_id in self.teams.keys()
        }

        winner_id = max(team_totals.items(), key=lambda x: x[1])[0]

        return {
            "team_id": winner_id,
            "team_name": self.teams[winner_id],
            "total_score": team_totals[winner_id]
        }

    def _display_final_results(self, results: Dict):
        """Display championship results"""
        print(f"\n{'='*60}")
        print("  FINAL RESULTS")
        print(f"{'='*60}\n")

        winner = results["winner"]
        print(f"🏆 WINNER: {winner['team_name']}")
        print(f"   Total Improvement: +{winner['total_score']:.1f} points\n")

        print("Recommended Flow Optimization Strategy:")
        print("  1. Optimize happy paths for conversion (Team 1)")
        print("  2. Handle edge cases and errors gracefully (Team 2)")
        print("  3. Ensure robust cross-flow integration (Team 3)")


def main():
    parser = argparse.ArgumentParser(description="User Flow Olympics")
    parser.add_argument(
        "--flows",
        nargs="+",
        required=True,
        help="Flows to test (e.g., registration login checkout)"
    )
    parser.add_argument("--rounds", type=int, default=4)
    parser.add_argument("--target", type=Path)

    args = parser.parse_args()

    coordinator = UserFlowCoordinator(
        flows=args.flows,
        rounds=args.rounds,
        target_path=args.target
    )
    coordinator.run_championship()


if __name__ == "__main__":
    main()

File: competitive-ai-frameworks/frameworks/user-flows/coordinator.ts
#!/usr/bin/env ts-node
/**
 * User Flow Testing Championship Coordinator
 *
 * Orchestrates competitive user flow testing between AI teams
 */

import * as fs from 'fs';
import * as path from 'path';

export interface UserFlow {
  flow_name: string;
  steps: string[];
  expected_outcome: string;
  test_status: 'pass' | 'fail' | 'pending';
}

export interface TeamConfig {
  name: string;
  strategy: string;
  focus_areas: string[];
  tools: string[];
}

export class UserFlowCoordinator {
  private target_path: string;
  private rounds: number;
  private output_dir: string;

  private teams: Record<string, TeamConfig>;
  private session_id: string;

  constructor(
    target_path: string,
    rounds: number = 5,
    output_dir: string = './results'
  ) {
    this.target_path = target_path;
    this.rounds = rounds;
    this.output_dir = output_dir;

    if (!fs.existsSync(this.output_dir)) {
      fs.mkdirSync(this.output_dir, { recursive: true });
    }
    this.session_id = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);

    this.teams = this._initialize_teams();
  }

  private _initialize_teams(): Record<string, TeamConfig> {
    return {
      team1: {
        name: 'Happy Path Optimizers',
        strategy: 'happy_path',
        focus_areas: ['common_workflows', 'user_expectations', 'optimal_paths'],
        tools: ['flow_analyzer', 'path_optimizer', 'usage_tracker'],
      },
      team2: {
        name: 'Edge Case Handlers',
        strategy: 'edge_cases',
        focus_areas: ['error_handling', 'boundary_conditions', 'unusual_inputs'],
        tools: ['edge_detector', 'error_simulator', 'boundary_tester'],
      },
      team3: {
        name: 'Integration Specialists',
        strategy: 'integration',
        focus_areas: ['cross_component', 'end_to_end', 'system_integration'],
        tools: ['integration_tester', 'e2e_framework', 'dependency_mapper'],
      },
    };
  }

  public async run_championship(): Promise<Record<string, any>> {
    console.log('\n' + '='.repeat(60));
    console.log('  USER FLOW TESTING CHAMPIONSHIP');
    console.log('='.repeat(60));
    console.log(`Target: ${this.target_path}`);
    console.log(`Rounds: ${this.rounds}`);
    console.log('='.repeat(60) + '\n');

    const results = {
      championship_id: this.session_id,
      target: this.target_path,
      rounds: this.rounds,
      teams: this.teams,
      message: 'User flow testing completed',
    };

    console.log('\n✓ Championship completed successfully\n');
    return results;
  }
}

// Main entry point
if (require.main === module) {
  const args = process.argv.slice(2);
  const target_path = args.find(arg => !arg.startsWith('--')) || process.cwd();
  const rounds = parseInt(args.find(arg => arg.startsWith('--rounds='))?.split('=')[1] || '5');
  const output_dir = args.find(arg => arg.startsWith('--output='))?.split('=')[1] || './results';

  if (!fs.existsSync(target_path)) {
    console.error(`Error: Target path does not exist: ${target_path}`);
    process.exit(1);
  }

  const coordinator = new UserFlowCoordinator(target_path, rounds, output_dir);
  coordinator
    .run_championship()
    .then(() => process.exit(0))
    .catch(error => {
      console.error('Error running championship:', error);
      process.exit(1);
    });
}
